### PR TITLE
 [Test] L4 complete diffusion feature test for Qwen-Image models

### PR DESCRIPTION
**Pull request by SamitHuang**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

This PR adds L4 test for Qwen-Image models, including Qwen-Image and Qwen-Image-2512 

Tests are picked up by nightly: test_*_expansion.py -m 'advanced_model and diffusion and H100'.

## Test Plan

The [diffusion acceleration doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/user_guide/diffusion_acceleration/) states that Qwen-Image and Qwen-Image-2512 support the full set of cache and parallelism features, except for HSDP/quantization. This test suite covers:

| Test ID | Feature | Server args (minimal for that feature) |
|---------|---------|----------------------------------------|
| `cache_tea_cache` | TeaCache | `--cache-backend tea_cache` (1 GPU) |
| `layerwise_offload` | CPU offload (layerwise) | `--cache-backend cache_dit --enable-layerwise-offload` (1 GPU) |
| `ulysses_2` | Ulysses-SP degree 2 | `--cache-backend cache_dit --ulysses-degree 2` (2 GPUs) |
| `ring_2` | Ring-Attention size 2 | `--cache-backend cache_dit --ring 2` (2 GPUs) |
| `cfg_parallel_2` | CFG-Parallel size 2 | `--cache-backend tea_cache --cfg-parallel-size 2` (2 GPUs) |
| `vae_patch_parallel_2` | VAE-Patch-Parallel size 2 | `--cache-backend cache_dit --tensor-parallel-size 2 --vae-patch-parallel-size 2 --vae-use-tiling` (2 GPUs; TP=2 is required to get 2 ranks for VAE patch) |

## Test Result
```
pytest tests/e2e/online_serving/test_qwen_image_expansion.py --collect-only -m diffusion
```

```
collected 12 items

<Dir vllm-omni>
  <Package tests>
    <Package e2e>
      <Package online_serving>
        <Module test_qwen_image_expansion.py>
          <Function test_qwen_image[cache_tea_cache]>
          <Function test_qwen_image_2512[cache_tea_cache]>
          <Function test_qwen_image[layerwise_offload]>
          <Function test_qwen_image_2512[layerwise_offload]>
          <Function test_qwen_image[ulysses_2]>
          <Function test_qwen_image_2512[ulysses_2]>
          <Function test_qwen_image[ring_2]>
          <Function test_qwen_image_2512[ring_2]>
          <Function test_qwen_image[cfg_parallel_2]>
          <Function test_qwen_image_2512[cfg_parallel_2]>
          <Function test_qwen_image[vae_patch_parallel_2]>
          <Function test_qwen_image_2512[vae_patch_parallel_2]>
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by wtomin** [comment]
@fhfuih PTAL

File: tests/e2e/online_serving/test_qwen_image_expansion.py
Comment (fhfuih) [inline tests/e2e/online_serving/test_qwen_image_expansion.py:1]: Regarding the test cases, plz
- Reduce the number of single-card tests. Only keep one tea_cache + layerwise_offload. ~~(cache dit and layerwise offloading is incompatible, pending PR 1786)~~ 1786 just merged. Can update this branch and keep this cache_dit + layerwise_offload combination, and only remove other single-card tests
- FP8 quantization should be supported. #1217 is sometimes more updated than the online documentation. And since GGUF is not supported, could you add FP8 to all test cases? (Unless it turns out to be incompatible with some certain features.)

Comment (fhfuih) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:1]: And also, we can combine tensor parallel and vae parallel into one test case to reduce time. Maybe you have referred to an earlier version of my PR before it is merged 😄 The current test cases in test_qwen_image_edit_expansion.py‎ is shorter

Comment (SamitHuang) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:1]: we still need to cover the tests for bf16 precision. let's add fp8 config to the largest combination only

Comment (fhfuih) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:1]: OK 👍 Then I may need to update relevant doc in another commit

**Comment by fhfuih** [comment]
Could you also force the current changes to run in "ready" CI pipeline, wait for it to success, and then change it back, with an additional comment to attach the result of that CI pipeline? Similar to 

https://github.com/vllm-project/vllm-omni/pull/1682/changes/aca42cb39fe11b2c0fc8a6fcfbff448fd106074e

https://github.com/vllm-project/vllm-omni/pull/1682#issuecomment-4029127978

Thanks!

**Comment by hsliuustc0106** [comment]
do we have an estimation for peak memory usage, latency for each scenario?

File: tests/e2e/online_serving/test_qwen_image_expansion.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                    "--cache-backend",
+                    "cache_dit",
+                    "--enable-layerwise-offload",
+                ],
+            ),
+            id="layerwise_offload",
+            marks=SINGLE_CARD_FEATURE_MARKS,
+        ),
+        pytest.param(
+            OmniServerParams(
```
Comment (hsliuustc0106) [inline tests/e2e/online_serving/test_qwen_image_expansion.py:49]: This test uses cache_dit for layerwise offload, while PR #1682 uses tea_cache. Is this intentional? If there's a known compatibility issue (referenced in #1779), please document it in the test. If both should behave the same, consider aligning the configuration.

File: tests/e2e/online_serving/test_qwen_image_expansion.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                    "--cfg-parallel-size",
+                    "2",
+                ],
+            ),
+            id="cfg_parallel_2",
+            marks=PARALLEL_FEATURE_MARKS,
+        ),
+        pytest.param(
+            OmniServerParams(
+                model=model,
```
Comment (hsliuustc0106) [inline tests/e2e/online_serving/test_qwen_image_expansion.py:89]: Why is FP8 quantization required for vae_patch_parallel_2? If it's mandatory for VAE-Patch-Parallel to work with TP=2, please add a comment explaining the constraint. If it's optional, consider removing it to simplify the test and isolate the feature being tested.

File: tests/e2e/online_serving/test_qwen_image_expansion.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
+"""
```
Comment (hsliuustc0106) [inline tests/e2e/online_serving/test_qwen_image_expansion.py:1]: This test suite covers each diffusion feature in isolation, but production deployments often combine multiple features (e.g., TeaCache + Ulysses, or layerwise offload with TP). Consider adding 1-2 combined feature tests to validate real-world scenarios. If this is intentionally out of scope, please document the limitation.

File: tests/e2e/online_serving/test_qwen_image_expansion.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        "extra_body": {
+            "height": 512,
+            "width": 512,
+            "num_inference_steps": 2,
+            "negative_prompt": NEGATIVE_PROMPT,
+            "true_cfg_scale": 4.0,
+            "seed": 42,
+        },
+    }
+    openai_client.send_diffusion_request(request_config)
```
Comment (hsliuustc0106) [inline tests/e2e/online_serving/test_qwen_image_expansion.py:130]: Verify that send_diffusion_request() provides clear failure messages for CI debugging. Specifically: (1) Does it check image dimensions match the request (512x512)? (2) Does it catch and report server errors with actionable information? If validation is minimal, consider adding explicit assertions.

File: tests/e2e/online_serving/test_qwen_image_expansion.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            id="ring_2",
+            marks=PARALLEL_FEATURE_MARKS,
+        ),
+        pytest.param(
+            OmniServerParams(
+                model=model,
+                server_args=[
+                    "--cache-backend",
+                    "tea_cache",
+                    "--cfg-parallel-size",
```
Comment (hsliuustc0106) [inline tests/e2e/online_serving/test_qwen_image_expansion.py:80]: The vae_patch_parallel_2 case combines 4 advanced features (TP=2, VAE-Patch-Parallel, tiling, FP8 quantization). This is the most complex test case and may be prone to flakiness in CI. Consider: (1) Adding a timeout to prevent hangs, (2) Logging which features are active when the test fails, (3) If this combination is rarely used in production, consider simplifying or documenting the edge-case nature.

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
Valuable test addition covering individual diffusion features. Key concerns:

1. **Configuration alignment**: layerwise_offload uses cache_dit here but tea_cache in PR #1682 - please document the difference or align.

2. **FP8 quantization**: Unclear why vae_patch_parallel_2 requires FP8. If mandatory, document it. If optional, remove it to isolate the feature.

3. **Combined features**: Missing tests for feature combinations (e.g., TeaCache + Ulysses). Consider adding 1-2 real-world scenarios.

4. **Error clarity**: Verify send_diffusion_request() provides actionable failure messages for CI debugging.

5. **Complexity risk**: vae_patch_parallel_2 combines 4 features (TP, VAE-Patch-Parallel, tiling, FP8) - may be flaky. Consider simplifying or adding timeouts/logging.

Structure follows established patterns well. Addressing these will improve maintainability and CI reliability.

**Comment by fhfuih** [comment]
> do we have an estimation for peak memory usage, latency for each scenario?

Not for this case, these tests mainly cover features.

**Comment by fhfuih** [comment]
> Valuable test addition covering individual diffusion features. Key concerns:
> 
> 1. **Configuration alignment**: layerwise_offload uses cache_dit here but tea_cache in PR #1682 - please document the difference or align.
> 2. **FP8 quantization**: Unclear why vae_patch_parallel_2 requires FP8. If mandatory, document it. If optional, remove it to isolate the feature.
> 3. **Combined features**: Missing tests for feature combinations (e.g., TeaCache + Ulysses). Consider adding 1-2 real-world scenarios.
> 4. **Error clarity**: Verify send_diffusion_request() provides actionable failure messages for CI debugging.
> 5. **Complexity risk**: vae_patch_parallel_2 combines 4 features (TP, VAE-Patch-Parallel, tiling, FP8) - may be flaky. Consider simplifying or adding timeouts/logging.
> 
> Structure follows established patterns well. Addressing these will improve maintainability and CI reliability.

Can confirm that the test cases are purposefully designed as such, to balance test duration and feature coverage with limited GPU resource. It has been discussed before.

File: tests/e2e/online_serving/test_qwen_image_expansion.py (hunk 2 above)
Comment (SamitHuang) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:89]: it's not required, just add it to test FP8 feature and whether it's compatable with other features including vae_patch_parallel_2

File: tests/e2e/online_serving/test_qwen_image_expansion.py (hunk 1 above)
Comment (SamitHuang) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:49]: it is because cache-dit is more widely used but there was some compatibility issues with other features

File: tests/e2e/online_serving/test_qwen_image_expansion.py (hunk 5 above)
Comment (SamitHuang) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:80]: it is mainly designed for covering more features in one test

File: tests/e2e/online_serving/test_qwen_image_expansion.py (hunk 3 above)
Comment (SamitHuang) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:1]: designed for limited computational resource. the largest combination covered cache + parallel

File: tests/e2e/online_serving/test_qwen_image_expansion.py (hunk 4 above)
Comment (fhfuih) [reply tests/e2e/online_serving/test_qwen_image_expansion.py:130]: The image shape verification is done within `send_diffusion_request`, which calls `assert_diffusion_response` function internally. See tests/conftets

**Comment by yenuo26** [comment]
Has this use case been debugged on CI? Are there any further modifications needed?

**Comment by Gaohan123** [comment]
Please supplement performance and accuracy baseline and check. If it is too heavy for all features, please add them for a recommended setting.
